classes/Hook/HookDisplayFooterProduct.php

```diff
@@ -42,73 +39,107 @@ public function __construct(Ps_Googleanalytics $module, Context $context)
-        $controllerName = Tools::getValue('controller');
-
-        if ('product' !== $controllerName) {
-            return '';
+        // Check we are really on product page
+        if ($this->context->controller->php_self !== 'product') {
+            return;
         }
 
-        if ($this->params['product'] instanceof Product) {
-            $this->params['product'] = (array) $this->params['product'];
+        // Get lazy array from context
+        $product = $this->context->smarty->getTemplateVars('product');
+        if (empty($product)) {
+            return;
         }
-        // Add product view
-        $js = $this->getGoogleAnalytics4();
 
-        return $gaTagHandler->generate($js);
-    }
+        // Initialize tag handler
+        $gaTagHandler = new GanalyticsJsHandler($this->module, $this->context);
 
-    /**
-     * setParams
-     *
-     * @param array $params
-     */
-    public function setParams($params)
-    {
-        $this->params = $params;
-    }
+        // Prepare it and format it for our purpose
+        $productWrapper = new ProductWrapper($this->context);
+        $item = $productWrapper->prepareItemFromProduct($product);
 
-    protected function getGoogleAnalytics4()
-    {
-        $gaProduct = $this->getProduct();
+        $jsCode = '';
+
+        // Prepare and render event
         $eventData = [
             'currency' => $this->context->currency->iso_code,
-            'value' => $this->params['product']['price_amount'],
-            'items' => [
-                [
-                    'item_id' => (int) $gaProduct['id'],
-                    'item_name' => $this->params['product']['name'],
-                    'currency' => $this->context->currency->iso_code,
-                    'item_brand' => $this->params['product']['manufacturer_name'],
-                    'item_category' => $this->params['product']['category_name'],
-                    'price' => (float) $this->params['product']['price_amount'],
-                    'quantity' => (int) $gaProduct['quantity'],
-                ],
-            ],
+            'value' => $item['price'],
+            'items' => [$item],
         ];
-        $js = $this->module->getTools()->renderEvent(
+        $jsCode .= $this->module->getTools()->renderEvent(
             'view_item',
             $eventData
         );
 
-        if (isset($_SERVER['HTTP_REFERER']) && strpos($_SERVER['HTTP_REFERER'], $_SERVER['HTTP_HOST']) > 0) {
-            $js .= $this->module->getTools()->addProductClickByHttpReferal([$gaProduct], $this->context->currency->iso_code);
+        // If the user got to the product page from previous page on our shop,
+        // we will also send select_item event.
+        if ($this->wasPreviousPageOurShop()) {
+            $eventData = [
+                'items' => [$item],
+            ];
+
+            // We will also try to get the information about the last visited listing.
```

Comment:
So many rules "if we are on page (...) then we do (...) but if we come from page (...) then we do (...)" it's hard to build a complete mental map of all the usecases.

I wonder if all this logic is in the right place in HookDisplayFooterProduct. How could we make it easier to know "if I am on hook (...) and doing (...) then data (...) will be sent" ?

Reply:
Yeah, but that's just how it's done, its million times clearer than the old code. :-)